File: sdk/go/status_reporter.go
package deploysentry

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"runtime/debug"
	"time"
)

// HealthReport is returned by a user-supplied health provider.
type HealthReport struct {
	State  string   `json:"state"` // healthy | degraded | unhealthy | unknown
	Score  *float64 `json:"score,omitempty"`
	Reason string   `json:"reason,omitempty"`
}

// versionEnvChain is probed in order; first non-empty wins.
var versionEnvChain = []string{
	"APP_VERSION",
	"GIT_SHA",
	"GIT_COMMIT",
	"SOURCE_COMMIT",
	"RAILWAY_GIT_COMMIT_SHA",
	"RENDER_GIT_COMMIT",
	"VERCEL_GIT_COMMIT_SHA",
	"HEROKU_SLUG_COMMIT",
}

const (
	defaultStatusInterval = 30 * time.Second
	statusMinBackoff      = 1 * time.Second
	statusMaxBackoff      = 5 * time.Minute
)

// resolveVersion picks the reported version from: explicit → env vars →
// build info → "unknown". Exposed for testing.
func resolveVersion(explicit string) string {
	if explicit != "" {
		return explicit
	}
	for _, name := range versionEnvChain {
		if v := os.Getenv(name); v != "" {
			return v
		}
	}
	if info, ok := debug.ReadBuildInfo(); ok && info.Main.Version != "" && info.Main.Version != "(devel)" {
		return info.Main.Version
	}
	return "unknown"
}

// statusReporter is owned by Client and fires periodic /status POSTs.
type statusReporter struct {
	client *Client
}

func newStatusReporter(c *Client) *statusReporter {
	return &statusReporter{client: c}
}

// run blocks until ctx is cancelled, firing reports on the configured cadence.
func (r *statusReporter) run(ctx context.Context) {
	// Initial report on start.
	r.report(ctx)

	interval := r.client.statusInterval
	if interval < 0 {
		interval = 0
	}
	if interval == 0 && r.client.statusInterval == 0 {
		// startup-only mode
		return
	}
	if interval == 0 {
		interval = defaultStatusInterval
	}

	backoff := time.Duration(0)
	timer := time.NewTimer(interval)
	defer timer.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-timer.C:
			if err := r.reportOnce(ctx); err != nil {
				r.client.logger.Printf("deploysentry: status report error: %v", err)
				if backoff == 0 {
					backoff = statusMinBackoff
				} else {
					backoff *= 2
					if backoff > statusMaxBackoff {
						backoff = statusMaxBackoff
					}
				}
				timer.Reset(backoff)
			} else {
				backoff = 0
				timer.Reset(interval)
			}
		}
	}
}

// report is the best-effort wrapper used during run() for non-test paths.
func (r *statusReporter) report(ctx context.Context) {
	if err := r.reportOnce(ctx); err != nil {
		r.client.logger.Printf("deploysentry: status report error: %v", err)
	}
}

// reportOnce sends exactly one POST. Exposed for tests and explicit triggers.
func (r *statusReporter) reportOnce(ctx context.Context) error {
	version := resolveVersion(r.client.statusVersion)
	health := HealthReport{State: "healthy"}
	if r.client.healthProvider != nil {
		got, err := r.client.healthProvider()
		if err != nil {
			health = HealthReport{State: "unknown", Reason: err.Error()}
		} else {
			health = got
		}
	}

	body := map[string]interface{}{
		"version": version,
		"health":  health.State,
	}
	if health.Score != nil {
		body["health_score"] = *health.Score
	}
	if health.Reason != "" {
		body["health_reason"] = health.Reason
	}
	if r.client.statusCommitSHA != "" {
		body["commit_sha"] = r.client.statusCommitSHA
	}
	if r.client.statusDeploySlot != "" {
		body["deploy_slot"] = r.client.statusDeploySlot
	}
	if len(r.client.statusTags) > 0 {
		body["tags"] = r.client.statusTags
	}

	buf, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("marshal status payload: %w", err)
	}

	endpoint := fmt.Sprintf("%s/api/v1/applications/%s/status", r.client.baseURL, url.PathEscape(r.client.applicationID))
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(buf))
	if err != nil {
		return fmt.Errorf("build status request: %w", err)
	}
	r.client.setAuthHeaders(req)
	req.Header.Set("Content-Type", "application/json")

	resp, err := r.client.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("post status: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("status report returned %s", resp.Status)
	}
	return nil
}

File: sdk/node/src/status-reporter.ts
import type { HealthReport } from './types';

/** Environment variable names probed (in order) for the running version. */
const VERSION_ENV_CHAIN = [
  'APP_VERSION',
  'GIT_SHA',
  'GIT_COMMIT',
  'SOURCE_COMMIT',
  'RAILWAY_GIT_COMMIT_SHA',
  'RENDER_GIT_COMMIT',
  'VERCEL_GIT_COMMIT_SHA',
  'HEROKU_SLUG_COMMIT',
] as const;

const MIN_BACKOFF_MS = 1_000;
const MAX_BACKOFF_MS = 5 * 60_000;

/** Options passed to StatusReporter. Internal to the SDK. */
export interface StatusReporterOptions {
  baseURL: string;
  apiKey: string;
  applicationId: string;
  intervalMs?: number;
  version?: string;
  commitSha?: string;
  deploySlot?: string;
  tags?: Record<string, string>;
  healthProvider?: () => HealthReport | Promise<HealthReport>;
  /** Injection seam for tests. Defaults to global fetch. */
  fetchImpl?: typeof fetch;
  /** Injection seam for tests. */
  warn?: (msg: string) => void;
}

/**
 * Resolve the version the app should report, in this preference order:
 * explicit config → standard env vars → `npm_package_version` → "unknown".
 */
export function resolveVersion(explicit?: string): string {
  if (explicit && explicit.trim()) return explicit;
  const env = typeof process !== 'undefined' && process.env ? process.env : {};
  for (const name of VERSION_ENV_CHAIN) {
    const v = env[name];
    if (v && v.trim()) return v;
  }
  if (env.npm_package_version) return env.npm_package_version;
  return 'unknown';
}

/**
 * StatusReporter posts periodic `POST /applications/:id/status` updates to
 * DeploySentry on behalf of a DeploySentryClient. Failures are swallowed
 * (logged via `warn`) so flag evaluation is never blocked by reporting.
 */
export class StatusReporter {
  private readonly opts: Required<
    Pick<StatusReporterOptions, 'baseURL' | 'apiKey' | 'applicationId' | 'intervalMs' | 'warn'>
  > &
    Omit<StatusReporterOptions, 'baseURL' | 'apiKey' | 'applicationId' | 'intervalMs' | 'warn'>;

  private timer: ReturnType<typeof setTimeout> | null = null;
  private stopped = false;
  private backoff = 0;

  constructor(options: StatusReporterOptions) {
    if (!options.baseURL) throw new Error('baseURL is required');
    if (!options.apiKey) throw new Error('apiKey is required');
    if (!options.applicationId) throw new Error('applicationId is required');
    const intervalMs = options.intervalMs ?? 30_000;
    if (intervalMs < 0) throw new Error('intervalMs must be >= 0');

    this.opts = {
      ...options,
      baseURL: options.baseURL.replace(/\/+$/, ''),
      apiKey: options.apiKey,
      applicationId: options.applicationId,
      intervalMs,
      warn: options.warn ?? ((msg) => console.warn('[DeploySentry] ' + msg)),
    };
  }

  /** Fire one immediate report and (if intervalMs > 0) schedule repeats. */
  start(): void {
    this.stopped = false;
    void this.tick();
  }

  /** Stop scheduling further reports. In-flight reports still resolve. */
  stop(): void {
    this.stopped = true;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }

  /** Send exactly one report. Public for tests and explicit callers. */
  async reportOnce(): Promise<void> {
    const version = resolveVersion(this.opts.version);
    let health: HealthReport = { state: 'healthy' };
    if (this.opts.healthProvider) {
      try {
        health = await this.opts.healthProvider();
      } catch (err) {
        health = { state: 'unknown', reason: String((err as Error)?.message ?? err) };
      }
    }

    const body: Record<string, unknown> = {
      version,
      health: health.state,
    };
    if (health.score !== undefined) body.health_score = health.score;
    if (health.reason) body.health_reason = health.reason;
    if (this.opts.commitSha) body.commit_sha = this.opts.commitSha;
    if (this.opts.deploySlot) body.deploy_slot = this.opts.deploySlot;
    if (this.opts.tags && Object.keys(this.opts.tags).length) body.tags = this.opts.tags;

    const fetchImpl = this.opts.fetchImpl ?? fetch;
    const url = `${this.opts.baseURL}/api/v1/applications/${encodeURIComponent(
      this.opts.applicationId,
    )}/status`;

    const res = await fetchImpl(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `ApiKey ${this.opts.apiKey}`,
      },
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      throw new Error(`status report failed: ${res.status} ${res.statusText}`);
    }
  }

  private async tick(): Promise<void> {
    if (this.stopped) return;
    try {
      await this.reportOnce();
      this.backoff = 0; // success — clear backoff
    } catch (err) {
      this.opts.warn(`status report error: ${(err as Error).message}`);
      if (this.backoff >= MAX_BACKOFF_MS) {
        // Already at the ceiling. Reset so the next schedule falls back to
        // intervalMs — otherwise a server that recovers mid-outage takes up
        // to MAX_BACKOFF_MS (5m) to be noticed, regardless of how tight the
        // operator configured intervalMs. On the next failure the 1s ladder
        // restarts, so the short-outage behavior (quickly re-probe while
        // the server is still down) is preserved.
        this.backoff = 0;
        this.opts.warn(
          `status reporter backoff reset; probing every ${this.opts.intervalMs}ms`,
        );
      } else {
        this.backoff = this.backoff === 0 ? MIN_BACKOFF_MS : Math.min(this.backoff * 2, MAX_BACKOFF_MS);
      }
    }
    if (this.stopped) return;
    if (this.opts.intervalMs === 0) return; // startup-only
    const nextDelay = this.backoff > 0 ? this.backoff : this.opts.intervalMs;
    this.timer = setTimeout(() => void this.tick(), nextDelay);
    // Allow Node to exit even when the timer is pending.
    if (typeof (this.timer as { unref?: () => void }).unref === 'function') {
      (this.timer as { unref?: () => void }).unref?.();
    }
  }
}
